klog.go

```diff
@@ -1290,10 +1295,70 @@ func (l *loggingT) createFiles(sev severity) error {
 const flushInterval = 5 * time.Second
 
 // flushDaemon periodically flushes the log file buffers.
-func (l *loggingT) flushDaemon() {
-	for range time.NewTicker(flushInterval).C {
-		l.lockAndFlushAll()
+type flushDaemon struct {
+	interval time.Duration
+	flush    func()
+	stopC    chan struct{}
+	stopDone chan struct{}
+}
+
+// newFlushDaemon retrun a new flushDaemon.
+func newFlushDaemon(interval time.Duration, flush func()) *flushDaemon {
+	return &flushDaemon{
+		interval: interval,
+		flush:    flush,
+	}
+}
+
+// run starts a goroutine that periodically calls the daemons flush function.
+// Calling run on an already running daemon will have no effect.
+func (f *flushDaemon) run() {
+	if f.isRunning() {
```

Comment:
Some more code and/or documentation is needed to ensure that this code is thread-safe.

As it stands now, the user might call `StopFlushDaemon` while the daemon is being started. Not very likely, but still...

We can probably require that `logging.mu` must be held when calling `run`, `stop` or `isRunning` and then lock that mutex in `StopFlushDaemon`.

Reply:
I'd prefer to lock the flushDaemon separate from the loggingT, by giving the daemon its own mutex. This way, flushDaemon can do the locking itself, which reduces complexity for the caller.

Reply:
I agree, this better. I was worried that `run` might be called frequently, but because `createFile` itself is under an if check, that won't be the case.